### PR TITLE
2025 03 25 evidence

### PR DESCRIPTION
**Pull request by thedavidmeister**
## Checks

By submitting this for review, I'm confirming I've done the following:
- [ ] made this PR as small as possible
- [ ] unit-tested any new functionality
- [ ] linked any relevant issues or PRs
- [ ] included screenshots (if this involves a front-end change)

## Summary by CodeRabbit

- **Documentation**
  - Introduced a guideline on evidence-based practices for product development, emphasizing automated methods for gathering evidence and setting objective benchmarks.
  - Added a document discussing evidence-based approaches in software development, promoting robust testing practices and strategies to capture a wide range of scenarios, including the importance of addressing edge cases.

### CODE REVIEW

File: principles/evidence.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+## Automate the production of evidence
+
+The number one reason for not producing evidence, is that it is often difficult
+to produce reliably.
+
+Automating the process of producing evidence has several benefits
+
+- Distributes access to the evidence across the entire team
+  - Reduces reliance on a single team member
+    - They may not be available in your time zone or due to other resonsibilities
```
Comment (thedavidmeister) [inline principles/evidence.md]:
```suggestion
    - They may not be available in your time zone or due to other responsibilities
```

File: principles/evidence.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  - Consider manual software testing, automated example based testing, and fuzz
+    testing. Each producing orders of magnitude more evidence in comparable time
+    frames than the last.
+- Standardises the evidence produced
+  - It's very difficult to directly compare inconsistently collected evidence
+    over a long period of time, undermining all the value produced
+
+Examples of evidence automation include:
+
+- Simple scripts that can be run on Github actions as needed
```
Comment (thedavidmeister) [inline principles/evidence.md]:
```suggestion
- Simple scripts that can be run on GitHub actions as needed
```

File: principles/evidence.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+## Establish a baseline before making large changes
+
+We are typically biased towards assuming that when we undertake some large
+project, the end result will be better than the current state of things.
+
+Indeed, why would we ever start a large new project if we expect a worse final
+outcome?
+
+The problem however, is often knowing when to stop, or at least to cut a 1.0
```
Comment (thedavidmeister) [inline principles/evidence.md]:
```suggestion
The problem, however, is often knowing when to stop, or at least to cut a 1.0
```

File: principles/evidence.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+We are typically biased towards assuming that when we undertake some large
+project, the end result will be better than the current state of things.
+
+Indeed, why would we ever start a large new project if we expect a worse final
+outcome?
+
+The problem, however, is often knowing when to stop, or at least to cut a 1.0
+release.
+
+Ideally we should release as early and often as possible, which would be as soon
```
Comment (thedavidmeister) [inline principles/evidence.md]:
```suggestion
Ideally, we should release as early and often as possible, which would be as soon
```

File: principles/evidence.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+And we actually need to do the work of establishing what the baseline is in the
+old setup, ideally before we even consider releasing the new one.
+
+Examples include:
+
+- "All the tests pass" for a codebase refactor
+- "Uptime and efficiency" for an arbitrage bot
+- "Volume and TVL" for a DEX
+- Etc.
```
Comment (thedavidmeister) [inline principles/evidence.md]:
```suggestion
- Etc.

```

File: principles/evidence/software-development.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+Consider Dijkstra's classic example of multiplying two numbers. We cannot
+directly test that `a * b = c` gives us the correct answer for all `a` and `b`
+for even a humble `u32` let alone the `uint256` beasts in Solidity.
+
+"Proof by exhaustion" is therefore not possible, so all we can do is try to
+explore as many representative regions of the solution space as we can, in the
+most efficient way possible.
+
+An example is the fuzzer in Foundry for testing Solidity code. Faced with
+basically every value having ~2^256 possible values, they generates values
```
Comment (thedavidmeister) [inline principles/evidence/software-development.md]:
```suggestion
basically every value having ~2^256 possible values, they generate values
```

File: principles/evidence/software-development.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+- The code does not exist
+- Code exists but does not compile
+- Code compiles but we've never seen it do what we want
+- We've seen the code do what we want once
+- We've seen the code do what we want several times
+- We can't produce any scenarios where the code doesn't do what we want
+
+Happy path thinking skips the last and very crucial step. The step where we
+actually create reproduceable scenarios (aka "tests") and exhaust ourselves and
```
Comment (thedavidmeister) [inline principles/evidence/software-development.md]:
```suggestion
actually create reproducible scenarios (aka "tests") and exhaust ourselves and
```

File: principles/evidence/software-development.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+Common culprits include
+
+- Failing to test error/edge cases
+- Failing to test different families of numerics
+    - `0`, `1`, plurals, negatives, infinites, NaNs, floats, etc.
+- Failing to test corrupt data, e.g. invalid utf-8 bytes in strings, etc.
+
+One of the main benefits of moving from example based testing to generative/fuzz
+testing is that deviation from the happy path is the default, making it much more
+difficult to get stuck in a false sense of security.
```
Comment (thedavidmeister) [inline principles/evidence/software-development.md]:
```suggestion
difficult to get stuck in a false sense of security.

```

**Review by hardyjosh — APPROVED** [review]
